### PR TITLE
Remove usage of slice_chars in script

### PR DESCRIPTION
**Pull request by SimonSapin**
It’s deprecated in the #6850 rustup.

The first commit changes some behavior which was previously incorrect: the spec says indices in DOM strings are UTF-16 code units, not `char` code points.

The second commit should not change behavior, unless I made a mistake.

r? @jdm

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6854)

### CODE REVIEW

**Comment by Ms2ger** [comment]
Needs tests

**Comment by SimonSapin** [comment]
I’ll tweak `tests/wpt/web-platform-tests/dom/nodes/CharacterData-*` to exercise these changes, but I don’t know what to do about `textinput`.

**Comment by jdm** [comment]
There are unit tests in test/unit/script/textinput.rs - can you add some there?

**Comment by SimonSapin** [comment]
I’ll look into that tomorrow

**Comment by SimonSapin** [comment]
Rebased and added some tests, including a new file marked `expected: CRASH` because of #6873.

**Comment by jdm** [comment]
So if a text input contains "🌠" with the cursor at the left side, what happens now if you press right and insert or delete a character? Can we write tests for that?

-S-awaiting-review +S-awaiting-answer +S-needs-code-changes  

---

Reviewed 2 of 2 files at r1, 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/textinput.rs, line 260 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-Jvrk8-FI8TkS40ejh40)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L260)):</sup>
s/adjusting/adjust/

---

<sup>**[components/script/textinput.rs, line 280 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjEebdLh2QTkW7DYz)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L280)):</sup>
adjust_selection_for_horizontal_change

---

<sup>**[components/script/textinput.rs, line 292 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjYUCWJIiJDQzYLAJ)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L292)):</sup>
This is a change in behaviour; previously performing a horizonal adjustment without pressing shift while a selection was present would warp the cursor to one end up of the selection, but no further. Do the tests in test/unit/script catch this?

---

<sup>**[components/script/textinput.rs, line 297 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjOLXCwimsNOBR1e_)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L297)):</sup>
perform_horizontal_adjustment

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6854)

**Comment by SimonSapin** [comment]
Added a couple tests for the scenarios described.

---

Review status: 7 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/textinput.rs, line 260 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-Jvrk8-FI8TkS40ejh40)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L260)):</sup>
Done.

---

<sup>**[components/script/textinput.rs, line 280 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjEebdLh2QTkW7DYz)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L280)):</sup>
Done.

---

<sup>**[components/script/textinput.rs, line 292 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjYUCWJIiJDQzYLAJ)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L292)):</sup>
Tests still pass with this code change, but I don’t understand how it is a behavior change. Could you explain a bit more?

---

<sup>**[components/script/textinput.rs, line 297 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjOLXCwimsNOBR1e_)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L297)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6854)

**Comment by SimonSapin** [comment]
Added the tests that I had apparently forgotten to commit.

**Comment by SimonSapin** [comment]
Rebased.

**Comment by jdm** [comment]
Reviewed 7 of 7 files at r3.
Review status: 9 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/textinput.rs, line 292 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6854#-JvrjYUCWJIiJDQzYLAJ)** ([raw file](https://github.com/servo/servo/blob/4daaff4a7322d565da742c27dd1f0073bb9a2370/components/script/textinput.rs#L292)):</sup>
When adjust_horizontal was called before this PR, if a selection existed but `select` was not `Selected`, we would return immediately to the caller after clearing the selection and updating the cursor position to either the beginning or ending of the previous selection. We now return from `adjust_selection_for_horizontal_change` and proceed to call `perform_horizontal adjustment`, which suggests that the edit point will end up one character before the selection start or one character after the selection end.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6854)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Review status: 9 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6854)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-squash 

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/textinput.rs, line 286 \[r4\]](https://reviewable.io:443/reviews/servo/servo/6854#-Jxa1C3LcLJprgoDcsRz)** ([raw file](https://github.com/servo/servo/blob/39e8662e6afea48edbd0a1a0969abe1623e211a3/components/script/textinput.rs#L286)):</sup>
nit: caret

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6854)

**Comment by jdm** [comment]
@bors-servo: retry
* #7367

File: tests/wpt/metadata/MANIFEST.json [hunk 1]
```diff
@@ -29106,6 +29106,12 @@
       "shadow-dom/shadow-trees/hosting-multiple-shadow-trees-005.html"
     ],
     "items": {},
+        ],
```
Comment (jdm) [inline tests/wpt/metadata/MANIFEST.json]: Rebase error?

**Comment by SimonSapin** [comment]
Rebased and ran `mach test-wpt --manifest-update` again, but some unrelated tests have been marked as "deleted" in `tests/wpt/metadata/MANIFEST.json`

r? for the changes in that file.

**Pull request merged**